PerfettoCds/Pipeline/CompositeDataCookers/PerfettoGenericEventCooker.cs

```diff
@@ -0,0 +1,120 @@
+        public IReadOnlyCollection<DataCookerPath> RequiredDataCookers => new[]
+        {
+            PerfettoPluginConstants.SliceCookerPath,
+            PerfettoPluginConstants.ArgCookerPath,
+            PerfettoPluginConstants.ThreadTrackCookerPath,
+            PerfettoPluginConstants.ThreadCookerPath,
+            PerfettoPluginConstants.ProcessCookerPath
+        };
+
+        [DataOutput]
+        public ProcessedEventData<PerfettoGenericEvent> GenericEvents { get; }
+
+        /// <summary>
+        /// The highest number of fields found in any single event.
+        /// </summary>
+        [DataOutput]
+        public int MaximumEventFieldCount { get; private set; }
+
+        public PerfettoGenericEventCooker() : base(PerfettoPluginConstants.GenericEventCookerPath)
+        {
+            this.GenericEvents =
+                new ProcessedEventData<PerfettoGenericEvent>();
+        }
+
+        public void OnDataAvailable(IDataExtensionRetrieval requiredData)
+        {
+            // Gather the data from all the SQL tables
+            var sliceData = requiredData.QueryOutput<ProcessedEventData<PerfettoSliceEvent>>(new DataOutputPath(PerfettoPluginConstants.SliceCookerPath, nameof(PerfettoSliceCooker.SliceEvents)));
+            var argData = requiredData.QueryOutput<ProcessedEventData<PerfettoArgEvent>>(new DataOutputPath(PerfettoPluginConstants.ArgCookerPath, nameof(PerfettoArgCooker.ArgEvents)));
+            var threadTrackData = requiredData.QueryOutput<ProcessedEventData<PerfettoThreadTrackEvent>>(new DataOutputPath(PerfettoPluginConstants.ThreadTrackCookerPath, nameof(PerfettoThreadTrackCooker.ThreadTrackEvents)));
+            var threadData = requiredData.QueryOutput<ProcessedEventData<PerfettoThreadEvent>>(new DataOutputPath(PerfettoPluginConstants.ThreadCookerPath, nameof(PerfettoThreadCooker.ThreadEvents)));
+            var processData = requiredData.QueryOutput<ProcessedEventData<PerfettoProcessEvent>>(new DataOutputPath(PerfettoPluginConstants.ProcessCookerPath, nameof(PerfettoProcessCooker.ProcessEvents)));
+
+            // Join them all together
+
+            // Slice data contains event name and a few more fields
+            // Arg data contains the debug annotations
+            // ThreadTrack data allows us to get to the thread
+            // Thread data gives us the thread name+ID and gets us the process
+            // Process data gives us the process name+ID
+            var joined = from slice in sliceData
+                         join arg in argData on slice.ArgSetId equals arg.ArgSetId into args
+                         join threadTrack in threadTrackData on slice.TrackId equals threadTrack.Id
+                         join thread in threadData on threadTrack.Utid equals thread.Utid
+                         join process in processData on thread.Upid equals process.Upid
+                         select new { slice, args, threadTrack, thread, process };
+
```

Comment:
What's your thoughts on just having another cooker that creates the generic events instead of querying all the other cookers?

Reply:
I see why this approach was taken, each event type requires a specific SQL query. 

Is there a "generic" query that could be performed across all data in Perfetto?
Wondering if a general query should be done at the trace processor level, then each cooker would compose the events

Reply:
@trgibeau and I were talking. His suggestion in English (in my words) is to do a basic trace load/query on trace load for data that every table needs. Then all the tables will have on-demand sql queries as they load. In this manner the trace loads faster, and some data in the trace is available to view. Only visible mini-graphs would attempt to load in this manner. Or if someone used a profile with say 2 tables only those would load not all the other tables. This can be saved for another PR

Reply:
I'll explain my cooker architecture and rationale a little.

In order to get all the data needed for a complete "Generic Event", I need to join data across multiple SQL tables (5 tables for GenericEvents). This join can happen in 2 places: across the trace_processor_shell interface or in the cookers. These SQL tables also contain data that is needed for other WPA tables that I will be adding in the future, like CPU scheduling, memory usage, etc.

So I was trying to follow the cooker principle that data should be processed as few times as possible and can be re-used by other composite cookers. Now if any WPA table needs data that comes from a SQL "slices" table, they can get that data from the Slice source cooker and do the join in LINQ. As opposed to making multiple join queries across the trace_processor_shell interface, where we have to query from the SQL slice table multiple times.

That was my thinking anyway and I haven't tested at scale yet so I don't know how this holds up. 

Once I start adding more tables there might be some bad delay and we might need to add some intelligence to which tables I query from and when. I'll schedule a meeting with the 3 of you to discuss the best cooker architecture once I start adding more WPA tables.